core/src/test/java/org/apache/brooklyn/enricher/stock/PercentageEnricherTest.java

```diff
@@ -0,0 +1,275 @@
+    }
+
+    @Test
+    public void currentNullTest() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, null);
+        app.sensors().set(totalSensor, 100d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, null);
+    }
+
+    @Test
+    public void totalNullTest() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, 50d);
+        app.sensors().set(totalSensor, null);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, null);
+    }
+
+    @Test
+    public void bothInputNullTest() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, null);
+        app.sensors().set(totalSensor, null);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, null);
+    }
+
+    @Test
+    public void currentZeroTest() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, 0d);
+        app.sensors().set(totalSensor, 100d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, 0d);
+    }
+
+    @Test
+    public void totalZeroTest() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, 50d);
+        app.sensors().set(totalSensor, 0d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, null);
+    }
+
+    @Test
+    public void totalLessThanCurrentTest() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, 50d);
+        app.sensors().set(totalSensor, 25d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, null);
+    }
+
+    @Test
+    public void oneHundredPercentTest() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, 50d);
+        app.sensors().set(totalSensor, 50d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, 100d);
+    }
+
+    @Test
+    public void negativeCurrent() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, -50d);
+        app.sensors().set(totalSensor, 100d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, null);
+    }
+
+    @Test
+    public void negativeTotal() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, 50d);
+        app.sensors().set(totalSensor, -100d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, null);
+    }
+
+    @Test
+    public void bothSourceNegative() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, -50d);
+        app.sensors().set(totalSensor, -100d);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, null);
+    }
+
+    @Test
+    public void totalDoubleMaxValue() {
+        addEnricher();
+
+        app.sensors().set(currentSensor, Double.MAX_VALUE);
+        app.sensors().set(totalSensor, Double.MAX_VALUE);
+        EntityAsserts.assertAttributeEqualsEventually(app, targetSensor, 100d);
+    }
+
+    //SETUP TESTS
+    @Test(expectedExceptions = IllegalArgumentException.class)
```

Comment:
Also want to test that it throws an exception if entity == producer && targetSensor == sourceSensor.